arangod/Aql/OptimizerRulePropagateConstantAttributes.cpp

```diff
@@ -0,0 +1,344 @@
+                  // don't remove equality lookups on shard keys, as this may
+                  // prevent the restrict-to-single-shard rule from being
+                  // applied later!
+                  return false;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+
+    LOG_RULE << "CHANGE MEMBER";
+    parentNode->changeMember(accessIndex, const_cast<AstNode*>(constantValue));
+    return true;
+  }
+
+  return false;
+}
+
+/// @brief traverses an AST part recursively and patches it by inserting
+/// constant values
+bool insertConstantAttributes(ExecutionPlan& plan, AstNode* node,
+                              std::size_t currentLimitCount,
+                              VarReplacementTransientMap const& replacements) {
+  if (node == nullptr) {
+    return false;
+  }
+
+  bool modified = false;
+
+  if (node->type == NODE_TYPE_OPERATOR_BINARY_AND) {
+    auto lhs = node->getMember(0);
+    auto rhs = node->getMember(1);
+
+    modified |=
+        insertConstantAttributes(plan, lhs, currentLimitCount, replacements);
+    modified |=
+        insertConstantAttributes(plan, rhs, currentLimitCount, replacements);
+  } else if (node->type == NODE_TYPE_OPERATOR_BINARY_EQ) {
+    auto lhs = node->getMember(0);
+    auto rhs = node->getMember(1);
+
+    if (!lhs->isConstant() && (rhs->type == NODE_TYPE_ATTRIBUTE_ACCESS ||
+                               rhs->type == NODE_TYPE_REFERENCE)) {
+      modified |= insertConstantAttribute(plan, node, 1, currentLimitCount,
+                                          replacements);
+    }
+    if (!rhs->isConstant() && (lhs->type == NODE_TYPE_ATTRIBUTE_ACCESS ||
+                               lhs->type == NODE_TYPE_REFERENCE)) {
+      modified |= insertConstantAttribute(plan, node, 0, currentLimitCount,
+                                          replacements);
+    }
+  }
+
+  return modified;
+}
+
+bool processQuery(ExecutionPlan& plan, ExecutionNode* root,
+                  VarReplacementMap replacements,
+                  std::size_t currentLimitCount) {
+  auto replacementsMutable = replacements.transient();
+
+  bool modified = false;
+
+  std::size_t localLimitCount = currentLimitCount;
+  // collect
+  for (auto* node = root->getSingleton(); node != nullptr;
+       node = node->getFirstParent()) {
+    LOG_RULE << "[" << node->id() << "] " << node->getTypeString();
+    if (node->getType() == EN::LIMIT) {
+      localLimitCount += 1;
+    }
+    if (node->getType() != EN::FILTER) {
+      continue;
+    }
+
+    auto fn = ExecutionNode::castTo<FilterNode const*>(node);
+    auto setter = plan.getVarSetBy(fn->inVariable()->id);
+    if (setter != nullptr && setter->getType() == EN::CALCULATION) {
+      auto cn = ExecutionNode::castTo<CalculationNode*>(setter);
+      auto expression = cn->expression();
+
+      collectConstantAttributes(const_cast<AstNode*>(expression->node()),
```

Comment:
Do we really have to const_cast away the constness of the node here? I would hope that `collectConstantAttributes` does not modify the node.